File: 백준/Gold/2470. 두 용액/두 용액.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
arr.sort()

l, r = 0, n - 1
tmp = abs(arr[l] + arr[r])
result = [arr[l], arr[r]]


while l < r:
    left = arr[l]
    right = arr[r]
    s = left + right
  
    if abs(s) < tmp:
        tmp = abs(s)
        result = [left, right]
        if tmp == 0:
          break
    if s < 0:
        l += 1
    else:
        r -= 1

print(result[0], result[1])

File: 백준/Gold/2470. 두 용액/두 용액.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		int[] arr = new int[n];

		StringTokenizer st = new StringTokenizer(br.readLine());
		for (int i = 0; i < n; i ++){
			arr[i] = Integer.parseInt(st.nextToken());
		}

		Arrays.sort(arr);
		int l = 0;
		int r = n - 1;
		int min = Integer.MAX_VALUE;
		int[] result = {0, 0};
		while (l < r){
			int sum = arr[l] + arr[r];
			if (Math.abs(sum) < min){
				min = Math.abs(sum);
				result[0] = arr[l];
				result[1] = arr[r];
				if (sum == 0){
					break;
				}
			}
			if (sum < 0)
				l+= 1;
			else
				r -= 1;
		}
		System.out.println(result[0] + " " + result[1]);
	}

}